### PR TITLE
Add BenchmarkBank.

### PR DESCRIPTION
**Pull request by petermattis**

### CODE REVIEW

**Comment by petermattis** [comment]
Note that this doesn't actually give reasonable output yet and perhaps points to one of the perf problems with `examples/sql_bank`. Run as:

```
go test -run - -bench BenchmarkBank -vmodule=bank_test=1
```

Fairly quickly, you'll see log messages like:

```
I0904 13:34:59.427848 25874 stopper.go:229  draining; tasks left:
2      kv/txn_coord_sender.go:466
```

That number takes seconds to drop to 0. Does that mean we're not cleaning up SQL transactions correctly? @tschottdorf, @vivekmenezes

**Comment by tschottdorf** [comment]
Possible. I'll take a look soon but want to point out that before we've
actually optimized the kv, the SQL benchmark isn't useful (yet). In
particular the batch stuff.

**Comment by petermattis** [comment]
I realize the KV layer isn't optimized yet, but this seems like a bug, not a performance issue.

**Comment by tschottdorf** [comment]
If we actually lose a transaction, it should take 10 seconds until the heartbeat goroutine gives up. Try changing `kv.defaultClientTimeout` to `time.Hour` or something like that. We should (hopefully) also have a test for this somewhere @vivekmenezes?

**Comment by petermattis** [comment]
Ok, changed `kv.defaultClientTimeout` to `time.Hour` and now the benchmark just hangs (presumably it will finish after an hour). I also sometimes see the following warning:

`W0904 14:40:02.107526 28811 sql/bank_test.go:142  txn "sql/server.go:104 sql" id=5b25290d key="\xff\n\x02\n\x01\tbank\x00\x01\n\x03" rw=true pri=64.89922230 iso=SERIALIZABLE stat=PENDING epo=0 ts=1441392001.942632675,0 orig=1441392001.942632675,0 max=1441392001.942632675,0 failed to push "sql/server.go:104 sql" id=f0b9b013 key="\xff\n\x02\n\x01\tbank\x00\x01\n\x03" rw=true pri=79.76041780 iso=SERIALIZABLE stat=PENDING epo=0 ts=1441392001.942632675,1 orig=1441392001.816409826,0 max=1441392001.816409826,0`

**Comment by vivekmenezes** [comment]
On a perhaps related subject while I was working on another change I see that 

go test ./sql/driver/... -count=100

fails. The failure is in TestOverlappingTransactions where txn2 fails when things slow down for whatever reason.

**Comment by vivekmenezes** [comment]
filed away https://github.com/cockroachdb/cockroach/issues/2366 to track that test failure.

File: sql/bank_test.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+			b.Fatal(err)
+		}
+
+		var placeholders bytes.Buffer
+		var values []interface{}
+		for i := 0; i < *numAccounts; i++ {
+			if i > 0 {
+				_, _ = placeholders.WriteString(", ")
+			}
+			fmt.Fprintf(&placeholders, "($%d, 0)", i+1)
```
Comment (vivekmenezes) [inline sql/bank_test.go]: just write the string out and then remove the first two bytes outside the loop at the end

File: sql/bank_test.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	}
+
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			from := rand.Intn(*numAccounts)
+			var to int
+			for {
+				to = rand.Intn(*numAccounts)
+				if from != to {
```
Comment (vivekmenezes) [inline sql/bank_test.go]: do you really care about this?

Comment (tamird) [inline sql/bank_test.go]: i think this is just rolling the dice to pick a random account

Comment (vivekmenezes) [inline sql/bank_test.go]: what I meant was
from, to := rand.Intn(*numAccounts), rand.Intn(*numAccounts)
if from == to {
continue
}

**Comment by vivekmenezes** [comment]
BTW there is one scenario under which the server will definitely hang, which is when the server looses a connection to the client. The connection close will not clean up the pending transaction. I doubt it's happening here but just wanted to mention it.

File: sql/bank_test.go (hunk 2 above)
Comment (tamird) [inline sql/bank_test.go]: that signals to the benchmark that an iteration was completed, which is wrong.

Comment (vivekmenezes) [inline sql/bank_test.go]: fair enough, if we want to be that precise.

**Comment by tschottdorf** [comment]
@petermattis ok, then it's presumably true that we're losing transactions. The Push failure in itself is expected sometimes, but it could be related to losing one (since it's connected to a txn restart).

@vivekmenezes do we have a test for this?

**Comment by tschottdorf** [comment]
logged #2368

File: sql/bank_test.go (hunk 1 above)
Comment (petermattis) [inline sql/bank_test.go]: I'm not sure exactly what you mean. Are you suggesting that I always add a `", "` prefix and then remove the first one? I don't see that as being any simpler than what is here now and I've used the above pattern elsewhere (e.g. in `sql/parser`).

File: sql/bank_test.go (hunk 2 above)
Comment (petermattis) [inline sql/bank_test.go]: As Tamir noted, I want each benchmark iteration (i.e. each call to `pb.Next()`) to perform one money movement.

**Comment by petermattis** [comment]
Can I get an LGTM? Or additional comments. Yeah, the benchmark doesn't give useful perf results (yet), but that's not the benchmark's fault.

**Comment by tamird** [comment]
LGTM

File: sql/bank_test.go (hunk 1 above)
Comment (vivekmenezes) [inline sql/bank_test.go]: its fine as is.

**Comment by vivekmenezes** [comment]
LGTM

**Pull request merged**